File: extras/localStreamNativeApp/python/test_native.py
#!/usr/bin/env python3
import json
import struct
import subprocess
import sys
import time

def test_native_app(message, script_path):
    print(f"Testing message: {message}")
    
    # Start the native app
    process = subprocess.Popen([
        sys.executable, script_path
    ], stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
    
    def read_native_message(stdout):
        """Read a message in native messaging format"""
        try:
            # Read 4-byte length
            raw_length = stdout.read(4)
            if len(raw_length) < 4:
                return None
            
            length = struct.unpack('=I', raw_length)[0]
            if length == 0:
                return None
                
            # Read the message
            raw_message = stdout.read(length)
            return json.loads(raw_message.decode('utf-8'))
        except:
            return None
    
    def send_native_message(stdin, message):
        """Send a message in native messaging format"""
        encoded = json.dumps(message).encode('utf-8')
        length = struct.pack('=I', len(encoded))
        stdin.write(length + encoded)
        stdin.flush()
    
    try:
        # Read startup message
        startup_msg = read_native_message(process.stdout)
        print(f"Startup message: {startup_msg}")
        
        # Send our test message
        send_native_message(process.stdin, message)
        
        # Read response
        response = read_native_message(process.stdout)
        print(f"Response: {response}")
        
        # Close stdin to signal end
        process.stdin.close()
        
        # Wait for process to complete
        process.wait(timeout=5)
        
        return response
        
    except Exception as e:
        print(f"Error: {e}")
        process.terminate()
        return None

if __name__ == "__main__":
    print("=== Testing Native App ===")
    
    # Test ping
    print("\n1. Testing ping:")
    test_native_app({"action": "ping"}, "native_app.py")
    
    # Test invalid action
    print("\n2. Testing invalid action:")
    test_native_app({"action": "invalid"}, "native_app.py")
    
    # Test get_stats
    print("\n3. Testing get_stats:")
    test_native_app({"action": "get_stats"}, "native_app.py")

    print("\n4. Testing store_tweet with new TweetData structure:")
    test_tweet_data = {
        "originator_id": "test_tweet_123",
        "timestamp": "2025-09-10T13:30:00.000Z",
        "type": "HomeTimeline",
        "data": {"text": "This is a test tweet", "user": {"screen_name": "test_user"}},
        "user_id": "user_456",
        "canSendToCA": True,
        "reason": None,
        "date_added": "2025-09-10T13:30:00.000Z"
    }
    test_native_app({"action": "store_tweet", "tweet": test_tweet_data}, "native_app.py")

File: extras/localStreamNativeApp/test_native.ts
#!/usr/bin/env bun
import { spawn } from "node:child_process";

interface TestMessage {
  action: string;
  tweet?: any;
  sql?: string;
  [key: string]: any;
}

async function testNativeApp(message: TestMessage, scriptPath: string): Promise<any> {
  console.log(`Testing message: ${JSON.stringify(message)}`);
  
  return new Promise((resolve, reject) => {
    // Start the native app
    const process = spawn('bun', ['run', scriptPath], {
      stdio: ['pipe', 'pipe', 'pipe']
    });
    
    let responseReceived = false;
    
    function readNativeMessage(stdout: NodeJS.ReadableStream): Promise<any> {
      return new Promise((resolve) => {
        const lengthBuffer = Buffer.alloc(4);
        let lengthBytesRead = 0;
        
        const readLength = () => {
          const chunk = stdout.read(4 - lengthBytesRead);
          if (chunk) {
            chunk.copy(lengthBuffer, lengthBytesRead);
            lengthBytesRead += chunk.length;
            
            if (lengthBytesRead === 4) {
              const messageLength = lengthBuffer.readUInt32LE(0);
              if (messageLength === 0) {
                resolve(null);
                return;
              }
              
              // Read the message
              const messageBuffer = Buffer.alloc(messageLength);
              let messageBytesRead = 0;
              
              const readMessage = () => {
                const chunk = stdout.read(messageLength - messageBytesRead);
                if (chunk) {
                  chunk.copy(messageBuffer, messageBytesRead);
                  messageBytesRead += chunk.length;
                  
                  if (messageBytesRead === messageLength) {
                    try {
                      const messageStr = messageBuffer.toString('utf-8');
                      resolve(JSON.parse(messageStr));
                    } catch (error) {
                      resolve(null);
                    }
                  } else {
                    stdout.once('readable', readMessage);
                  }
                } else {
                  stdout.once('readable', readMessage);
                }
              };
              
              readMessage();
            } else {
              stdout.once('readable', readLength);
            }
          } else {
            stdout.once('readable', readLength);
          }
        };
        
        stdout.once('readable', readLength);
      });
    }
    
    function sendNativeMessage(stdin: NodeJS.WritableStream, message: TestMessage): void {
      const messageStr = JSON.stringify(message);
      const messageBuffer = Buffer.from(messageStr, 'utf-8');
      const lengthBuffer = Buffer.alloc(4);
      lengthBuffer.writeUInt32LE(messageBuffer.length, 0);
      
      stdin.write(lengthBuffer);
      stdin.write(messageBuffer);
    }
    
    // Handle process events
    process.on('error', (error) => {
      console.log(`Process error: ${error}`);
      reject(error);
    });
    
    process.on('close', (code) => {
      if (!responseReceived) {
        console.log(`Process closed with code: ${code}`);
        resolve(null);
      }
    });
    
    // Wait for startup message then send test message
    readNativeMessage(process.stdout!)
      .then(startupMessage => {
        console.log(`Startup message: ${JSON.stringify(startupMessage)}`);
        
        // Send our test message
        sendNativeMessage(process.stdin!, message);
        
        // Read response
        return readNativeMessage(process.stdout!);
      })
      .then(response => {
        console.log(`Response: ${JSON.stringify(response)}`);
        responseReceived = true;
        
        // Close stdin to signal end
        process.stdin!.end();
        
        resolve(response);
      })
      .catch(error => {
        console.log(`Error: ${error}`);
        process.kill();
        reject(error);
      });
  });
}

async function runTests() {
  console.log("=== Testing Bun Native App ===");
  
  try {
    // Test ping
    console.log("\n1. Testing ping:");
    await testNativeApp({ action: "ping" }, "native_app.ts");
    
    // Test invalid action  
    console.log("\n2. Testing invalid action:");
    await testNativeApp({ action: "invalid" }, "native_app.ts");
    
    // Test get_stats
    console.log("\n3. Testing get_stats:");
    await testNativeApp({ action: "get_stats" }, "native_app.ts");
    
    // Test store_tweet with new TweetData structure
    console.log("\n4. Testing store_tweet with new TweetData structure:");
    const testTweetData = {
      originator_id: "test_tweet_123",
      timestamp: "2025-09-10T13:30:00.000Z",
      type: "HomeTimeline",
      data: { text: "This is a test tweet", user: { screen_name: "test_user" } },
      user_id: "user_456",
      canSendToCA: true,
      reason: null,
      date_added: "2025-09-10T13:30:00.000Z"
    };
    await testNativeApp({ action: "store_tweet", tweet: testTweetData }, "native_app.ts");
    
  } catch (error) {
    console.error("Test failed:", error);
  }
}

runTests();
